File: C#/Silverlight/source/LSilverlight-0.3.3/Loon.Utils.Xml/XMLTokenizer.cs
namespace Loon.Utils.Xml {
	
	public class XMLTokenizer {
	
		private string text;
	
		private int pointer;
	
		public XMLTokenizer(string t) {
			this.text = t;
		}
	
		public bool HasMoreElements() {
			return this.pointer < this.text.Length;
		}
	
		public string NextElement() {
			if (this.text[this.pointer] == '<') {
				return NextTag();
			}
			return NextString();
		}
	
		private string NextTag() {
			int i = 0;
			int j = this.pointer;
			do {
				switch ((int) this.text[this.pointer]) {
				case '"':
					i = (i != 0) ? 0 : 1;
					break;
				}
				this.pointer += 1;
			} while ((this.pointer < this.text.Length)
					&& ((this.text[this.pointer] != '>') || (i != 0)));
			if (this.pointer < this.text.Length) {
				this.pointer += 1;
			} else {
				throw new System.Exception(
						"Tokenizer error: < without > at end of text");
			}
			return this.text.Substring(j,(this.pointer)-(j));
		}
	
		private string NextString() {
			int i = 0;
			int j = this.pointer;
			do {
				switch ((int) this.text[this.pointer]) {
				case '"':
					i = (i != 0) ? 0 : 1;
					break;
				}
				this.pointer += 1;
			} while ((this.pointer < this.text.Length)
					&& ((this.text[this.pointer] != '<') || (i != 0)));
			return this.text.Substring(j,(this.pointer)-(j));
		}
	
	}
}

File: Java/JavaSE/OpenGL/source/core/src/loon/utils/xml/XMLTokenizer.java
package loon.utils.xml;

/**
 * Copyright 2008 - 2011
 * 
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not
 * use this file except in compliance with the License. You may obtain a copy of
 * the License at
 * 
 * http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations under
 * the License.
 * 
 * @project loon
 * @author cping
 * @email：javachenpeng@yahoo.com
 * @version 0.1
 */
public class XMLTokenizer {

	private String text;

	private int pointer;

	public XMLTokenizer(String text) {
		this.text = text;
	}

	public boolean hasMoreElements() {
		return this.pointer < this.text.length();
	}

	public String nextElement() {
		if (this.text.charAt(this.pointer) == '<') {
			return nextTag();
		}
		return nextString();
	}

	private String nextTag() {
		int i = 0;
		int j = this.pointer;
		do {
			switch (this.text.charAt(this.pointer)) {
			case '"':
				i = i != 0 ? 0 : 1;
			}
			this.pointer += 1;
		} while ((this.pointer < this.text.length())
				&& ((this.text.charAt(this.pointer) != '>') || (i != 0)));
		if (this.pointer < this.text.length()) {
			this.pointer += 1;
		} else {
			throw new RuntimeException(
					"Tokenizer error: < without > at end of text");
		}
		return this.text.substring(j, this.pointer);
	}

	private String nextString() {
		int i = 0;
		int j = this.pointer;
		do {
			switch (this.text.charAt(this.pointer)) {
			case '"':
				i = i != 0 ? 0 : 1;
			}
			this.pointer += 1;
		} while ((this.pointer < this.text.length())
				&& ((this.text.charAt(this.pointer) != '<') || (i != 0)));
		return this.text.substring(j, this.pointer);
	}

}
